File: car/cluster/src/main/java/androidx/car/cluster/navigation/Distance.java
/*
 * Copyright 2018 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.car.cluster.navigation;

import static androidx.annotation.RestrictTo.Scope.LIBRARY_GROUP_PREFIX;

import android.annotation.SuppressLint;

import androidx.annotation.NonNull;
import androidx.annotation.RestrictTo;
import androidx.core.util.Preconditions;
import androidx.versionedparcelable.ParcelField;
import androidx.versionedparcelable.VersionedParcelable;
import androidx.versionedparcelable.VersionedParcelize;

import java.util.Objects;

/**
 * Distance along the planned route between relevant points in the navigation
 */
@VersionedParcelize
public final class Distance implements VersionedParcelable {
    /**
     * Possible units used to display this distance.
     */
    public enum Unit {
        /**
         * Display unit is unknown to the consumer, in which case, the distance shouldn't be
         * displayed.
         */
        UNKNOWN,
        METERS,
        KILOMETERS,
        MILES,
        FEET,
        YARDS,
    }

    @ParcelField(1)
    int mMeters;
    @ParcelField(2)
    String mDisplayValue = "";
    @ParcelField(3)
    EnumWrapper<Unit> mDisplayUnit;

    /**
     * Used by {@link VersionedParcelable}

     * @hide
     */
    @RestrictTo(LIBRARY_GROUP_PREFIX)
    Distance() {
    }

    /**
     * Creates a distance.
     *
     * @param meters distance in meters.
     * @param displayValue distance measured in {@code displayUnit}.
     * @param displayUnit unit to be used when displaying this distance
     */
    public Distance(int meters, @NonNull String displayValue,
            @NonNull Unit displayUnit) {
        mMeters = meters;
        mDisplayValue = Preconditions.checkNotNull(displayValue);
        mDisplayUnit = EnumWrapper.of(displayUnit);
    }

    /**
     * Returns the distance in meters.
     */
    public int getMeters() {
        return mMeters;
    }

    /**
     * Returns the distance measured in the unit indicated at {@link #getDisplayUnit()}, already
     * internationalized and ready for display, or empty if not distance value was provided. In the
     * later case, the consumer shouldn't display this distance to the driver.
     * <p>
     * This distance is for display only (it might be a rounded representation of the actual
     * distance) and it should mirror the distance displayed by the producer in its own UI.
     * <p>
     * For example, a distance of 1.2 km in {@code ES_es} locale would be represented as {@code
     * displayValue = "1,2"} and {@code displayUnit = KILOMETERS}). This field is only relevant if
     * {@link #getDisplayUnit()} is not {@link Unit#UNKNOWN}.
     */
    @NonNull
    public String getDisplayValue() {
        return Common.nonNullOrEmpty(mDisplayValue);
    }

    /**
     * Returns the distance unit (adjusted to the current user's locale and/or location). This field
     * would mirror the distance unit displayed by the producer in its own UI, and it should be
     * used for display only (no unit transformation should be applied).
     */
    @NonNull
    public Unit getDisplayUnit() {
        return EnumWrapper.getValue(mDisplayUnit, Unit.UNKNOWN);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        Distance distance = (Distance) o;
        return getMeters() == distance.getMeters()
                && Objects.equals(getDisplayValue(), distance.getDisplayValue())
                && Objects.equals(getDisplayUnit(), distance.getDisplayUnit());
    }

    @Override
    public int hashCode() {
        return Objects.hash(getMeters(), getDisplayUnit(), getDisplayValue());
    }

    // DefaultLocale suppressed as this method is only offered for debugging purposes.
    @SuppressLint("DefaultLocale")
    @Override
    public String toString() {
        return String.format("{meters: %d, displayUnit: %s, displayValue: %s}", mMeters,
                mDisplayUnit, mDisplayValue);
    }
}

File: health/health-data-client/src/main/java/androidx/health/data/client/records/Distance.kt
/*
 * Copyright (C) 2022 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package androidx.health.data.client.records

import androidx.health.data.client.aggregate.DoubleAggregateMetric
import androidx.health.data.client.metadata.Metadata
import java.time.Instant
import java.time.ZoneOffset

/**
 * Captures distance travelled by the user since the last reading, in meters. The total distance
 * over an interval can be calculated by adding together all the values during the interval. The
 * start time of each record should represent the start of the interval in which the distance was
 * covered.
 *
 * If break downs are preferred in scenario of a long workout, consider writing multiple distance
 * records. The start time of each record should be equal to or greater than the end time of the
 * previous record.
 */
public class Distance(
    /** Distance in meters. Required field. Valid range: 0-1000000. */
    public val distanceMeters: Double,
    override val startTime: Instant,
    override val startZoneOffset: ZoneOffset?,
    override val endTime: Instant,
    override val endZoneOffset: ZoneOffset?,
    override val metadata: Metadata = Metadata.EMPTY,
) : IntervalRecord {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Distance) return false

        if (distanceMeters != other.distanceMeters) return false
        if (startTime != other.startTime) return false
        if (startZoneOffset != other.startZoneOffset) return false
        if (endTime != other.endTime) return false
        if (endZoneOffset != other.endZoneOffset) return false
        if (metadata != other.metadata) return false

        return true
    }

    override fun hashCode(): Int {
        var result = 0
        result = 31 * result + distanceMeters.hashCode()
        result = 31 * result + (startZoneOffset?.hashCode() ?: 0)
        result = 31 * result + endTime.hashCode()
        result = 31 * result + (endZoneOffset?.hashCode() ?: 0)
        result = 31 * result + metadata.hashCode()
        return result
    }

    internal companion object {
        /** Metric identifier to retrieve total distance from [AggregateDataRow]. */
        @JvmStatic
        internal val DISTANCE_TOTAL: DoubleAggregateMetric =
            DoubleAggregateMetric("Distance", "total", "distance")
    }
}
